apiserver/facades/controller/charmrevisionupdater/updater_test.go

```diff
@@ -4,136 +4,126 @@
-		Charm:      ch,
-		ForceUnits: true,
-	}
-	err = app.SetCharm(cfg)
+	state.EXPECT().AllApplications().Return([]charmrevisionupdater.Application{
+		makeApplication(ctrl, "ch", "varnish", "charm-5", "app-1", 1),
+		makeApplication(ctrl, "cs", "varnish", "charm-6", "app-2", 2),
+	}, nil).AnyTimes()
+
+	updater, err := charmrevisionupdater.NewCharmRevisionUpdaterAPIState(state, newFakeCharmstoreClient, newFakeCharmhubClient)
 	c.Assert(err, jc.ErrorIsNil)
 
-	result, err = s.charmrevisionupdater.UpdateLatestRevisions()
+	result, err := updater.UpdateLatestRevisions()
 	c.Assert(err, jc.ErrorIsNil)
 	c.Assert(result.Error, gc.IsNil)
-
-	// Latest mysql is now deployed, so no pending charm.
-	curl = charm.MustParseURL("cs:quantal/mysql")
-	_, err = s.State.LatestPlaceholderCharm(curl)
-	c.Assert(err, jc.Satisfies, errors.IsNotFound)
 }
 
-func (s *charmVersionSuite) TestWordpressCharmNoReadAccessIsNotVisible(c *gc.C) {
-	s.AddMachine(c, "0", state.JobManageModel)
-	s.SetupScenario(c)
+func (s *updaterSuite) TestCharmstoreUpdate(c *gc.C) {
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
+
+	state := makeState(c, ctrl, nil)
+
+	state.EXPECT().AllApplications().Return([]charmrevisionupdater.Application{
+		makeApplication(ctrl, "cs", "mysql", "charm-1", "app-1", 22),
+		makeApplication(ctrl, "cs", "wordpress", "charm-2", "app-2", 26),
+		makeApplication(ctrl, "cs", "varnish", "charm-3", "app-3", 5), // doesn't exist in store
+	}, nil)
 
-	// Disallow read access to the wordpress charm in the charm store.
-	s.CharmSuite.SetStoreError("wordpress", errors.New("boom"))
+	state.EXPECT().AddCharmPlaceholder(charm.MustParseURL("cs:mysql-23")).Return(nil)
+	state.EXPECT().AddCharmPlaceholder(charm.MustParseURL("cs:wordpress-26")).Return(nil)
 
-	// Run the revision updater and check that the public charm updates are
-	// still properly notified.
-	result, err := s.charmrevisionupdater.UpdateLatestRevisions()
+	updater, err := charmrevisionupdater.NewCharmRevisionUpdaterAPIState(state, newFakeCharmstoreClient, nil)
 	c.Assert(err, jc.ErrorIsNil)
-	c.Assert(result.Error, gc.IsNil)
 
-	curl := charm.MustParseURL("cs:quantal/mysql")
-	pending, err := s.State.LatestPlaceholderCharm(curl)
+	result, err := updater.UpdateLatestRevisions()
 	c.Assert(err, jc.ErrorIsNil)
-	c.Assert(pending.String(), gc.Equals, "cs:quantal/mysql-23")
+	c.Assert(result.Error, gc.IsNil)
+
+	// Update mysql version and run update again.
+	state.EXPECT().AllApplications().Return([]charmrevisionupdater.Application{
+		makeApplication(ctrl, "cs", "mysql", "charm-1", "app-1", 23),
+	}, nil)
 
-	// No pending charm for wordpress.
-	curl = charm.MustParseURL("cs:quantal/wordpress")
-	_, err = s.State.LatestPlaceholderCharm(curl)
-	c.Assert(err, jc.Satisfies, errors.IsNotFound)
+	state.EXPECT().AddCharmPlaceholder(charm.MustParseURL("cs:mysql-23")).Return(nil)
+
+	result, err = updater.UpdateLatestRevisions()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(result.Error, gc.IsNil)
```

Comment:
Same here - the old code could check for no pending charms. What can we check for now?

Reply:
As above